daemon/src/rollover/taker.rs

```diff
@@ -0,0 +1,223 @@
+            tasks: Tasks::default(),
+            executor,
+            get_announcement,
+            oracle_pk,
+            n_payouts,
+        }
+    }
+}
+
+#[xtra_productivity]
+impl Actor {
+    pub async fn handle(&mut self, msg: ProposeRollover) {
+        let ProposeRollover {
+            order_id,
+            maker_peer_id,
+        } = msg;
+
+        let substream = match self
+            .endpoint
+            .send(OpenSubstream::single_protocol(
+                maker_peer_id.inner(),
+                rollover::PROTOCOL,
+            ))
+            .await
+            .context("Endpoint is disconnected")
+            .context("Failed to open substream")
+        {
+            Ok(Ok(substream)) => substream,
+            // TODO: Same error path and add failed
+            Ok(Err(e)) => {
+                tracing::error!("Failed to open connection to maker: {e:#}");
+                return;
+            }
+            Err(e) => {
+                tracing::error!(%order_id, "Failed to open connection to maker for  rollover: {e:#}");
+
+                if let Err(e) = self
+                    .executor
+                    .execute(order_id, |cfd| Ok(cfd.fail_rollover(e)))
+                    .await
+                {
+                    tracing::error!(%order_id, "Failed to execute rollover failed: {e:#}")
+                }
+
+                return;
+            }
+        };
+
+        self.tasks.add_fallible(
+            {
+                let executor = self.executor.clone();
+                let get_announcement = self.get_announcement.clone_channel();
+                let oracle_pk = self.oracle_pk;
+                let n_payouts = self.n_payouts;
+                async move {
+                    let mut framed = asynchronous_codec::Framed::new(
+                        substream,
+                        asynchronous_codec::JsonCodec::<DialerMessage, ListenerMessage>::new(),
+                    );
+
+                    executor
+                        .execute(order_id, |cfd| cfd.start_rollover())
+                        .await?;
+
+                    framed
+                        .send(DialerMessage::Propose(Propose {
+                            order_id,
+                            timestamp: Timestamp::now(),
+                        }))
+                        .await
+                        .context("Failed to send Msg0")?;
+
+                    match framed
+                        .next()
+                        .timeout(DECISION_TIMEOUT)
+                        .await
+                        .with_context(|| {
+                            format!(
+                                "Maker did not accept/reject within {} seconds.",
+                                DECISION_TIMEOUT.as_secs()
+                            )
+                        })?
+                        .context("End of stream while receiving Msg1")?
+                        .context("Failed to decode Msg1")?
```

Comment:
🔧 Probably should use something more descriptive than `Msg1` in the error message.